pilot/pkg/model/gateway_test.go

```diff
@@ -0,0 +1,268 @@
+				{
+					Port:  port80,
+					Tls:   tlsOne,
+					Hosts: []string{"bar.com", "foo.com"},
+				},
+			}},
+			nil,
+		},
+		{"different domains, different ports",
+			&routing.Gateway{Servers: []*routing.Server{
+				{
+					Port:  port80,
+					Tls:   tlsOne,
+					Hosts: []string{"foo.com"},
+				},
+			}},
+			&routing.Gateway{Servers: []*routing.Server{
+				{
+					Port:  port443,
+					Tls:   tlsTwo,
+					Hosts: []string{"bar.com"},
+				},
+			}},
+			&routing.Gateway{Servers: []*routing.Server{
+				{
+					Port:  port80,
+					Tls:   tlsOne,
+					Hosts: []string{"foo.com"},
+				}, {
+					Port:  port443,
+					Tls:   tlsTwo,
+					Hosts: []string{"bar.com"},
+				}}},
+			nil,
+		},
+		{"conflicting port names",
+			&routing.Gateway{Servers: []*routing.Server{
+				{
+					Port:  port80,
+					Tls:   tlsOne,
+					Hosts: []string{"foo.com"},
+				},
+			}},
+			&routing.Gateway{Servers: []*routing.Server{
+				{
+					Port:  port80DifferentName,
+					Tls:   tlsOne,
+					Hosts: []string{"bar.com"},
+				},
+			}},
+			nil,
+			errors.New(`unable to merge gateways: conflicting ports`),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// we want to save the original state of tt.a for printing if we fail the test, so we'll merge into a new gateway struct.
+			actual := &routing.Gateway{}
+			MergeGateways(actual, tt.a)
+			err := MergeGateways(actual, tt.b)
+			if err != tt.expectedErrPrefix && !strings.HasPrefix(err.Error(), tt.expectedErrPrefix.Error()) {
+				t.Fatalf("%s: got err %v, wanted %v", tt.name, err, tt.expectedErrPrefix)
+			}
+			if err == nil {
+				if !reflect.DeepEqual(actual, tt.expectedOut) {
+					t.Fatalf("%s: got %v, wanted %v", tt.name, actual, tt.expectedOut)
+				}
+			}
+		})
+	}
+
+}
+
+func TestPortsEqualAndDistinct(t *testing.T) {
+	tests := []struct {
+		name             string
+		b                *routing.Port
+		a                *routing.Port
+		expectedEqual    bool
+		expectedDistinct bool
+	}{
+		{"empty", &routing.Port{}, &routing.Port{}, true, false},
+		{"happy",
+			&routing.Port{Number: 1, Name: "Bill", Protocol: "HTTP"},
+			&routing.Port{Number: 1, Name: "Bill", Protocol: "HTTP"},
+			true, false},
+		{"same numbers but different names (case sensitive)",
+			&routing.Port{Number: 1, Name: "Bill", Protocol: "HTTP"},
+			&routing.Port{Number: 1, Name: "bill", Protocol: "HTTP"},
+			false, false},
+		{"case insensitive",
+			&routing.Port{Number: 1, Name: "potato", Protocol: "GRPC"},
+			&routing.Port{Number: 1, Name: "potato", Protocol: "grpc"},
+			true, false},
+		{"different numbers but same names",
+			&routing.Port{Number: 1, Name: "potato", Protocol: "tcp"},
+			&routing.Port{Number: 2, Name: "potato", Protocol: "tcp"},
+			false, false},
+		{"different protocols but same names",
+			&routing.Port{Number: 1, Name: "potato", Protocol: "http2"},
+			&routing.Port{Number: 1, Name: "potato", Protocol: "http"},
+			false, false},
```

Comment:
What is expectedDistinct ? Isn’t it not equal?

Reply:
Btw, please add a todo or file a follow up issue. We should consider h1/h2 as same port. Essentially all family of http protocols under http umbrella.